### ISSUE TITLE
Souporcell doublet detection fails on tutorial data (troublet step)

### ISSUE THREAD

**Issue by mschilli87** [drneavin/Demultiplexing_Doublet_Detecting_Docs#39]
Using [v2.1.0 of the singularity container](https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/releases/tag/v2.1.0) and [the reduced test dataset](https://demultiplexing-doublet-detecting-docs.readthedocs.io/en/latest/DataPrep.html#testdata), I fail to run Souporcell (without refence SNP genotypes):

```sh
souporcell_pipeline.py \
  -i test_dataset/outs/pooled.sorted.bam \
  -b test_dataset/outs/filtered_gene_bc_matrices/Homo_sapiens_GRCh38p10/barcodes.tsv \
  -f annotations/FASTA/hg38.fa \
  -t 32 \
  -o output/souporcell \
  -k 14 \
  --common_variants test_dataset.vcf
```

runs fine up to the

```
...
running vartrix
running souporcell clustering
...
```

steps but then

```
...
running souporcell doublet detection
...
```

fails with the following error:

```
Traceback (most recent call last):
  File "/opt/souporcell/souporcell_pipeline.py", line 598, in <module>
    doublets(args, ref_mtx, alt_mtx, cluster_file)
  File "/opt/souporcell/souporcell_pipeline.py", line 543, in doublets
    subprocess.check_call([directory+"/troublet/target/release/troublet", "--alts", alt_mtx, "--refs", ref_mtx, "--clusters", cluster_file], stdout = dub, stderr = err)
  File "/opt/conda/envs/py36/lib/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/souporcell/troublet/target/release/troublet', '--alts', 'output/souporcell/alt.mtx', '--refs', 'output/souporcell/ref.mtx', '--clusters', 'output/souporcell/clusters_tmp.tsv']' returned non-zero exit status 101.
```

Any idea what might be going wrong or advice on how to trouble-shoot this issue further?

**Comment by drneavin** [comment]
Hi @mschilli87 , can you check the contents of the alt.mtx and ref.mtx files? I typically see something like this when there has been a problem with those files that results in no data in them

**Comment by mschilli87** [comment]
@drneavin: Looks like we are on the right track:

```
%%MatrixMarket matrix coordinate real general
% written by sprs
0 1400 0
```

Can you confirm that running the my command on the mentioned data works for you?

**Comment by drneavin** [comment]
Hi @mschilli87 , yes, I just retested the image with the full dataset (I usually test on the small dataset of 1000 cells) and it ran completely. I previously have run into this issue if I ran out of memory while running souporcell and restarted it or if the reference didn't use the same chr encoding (chr1, chr2 vs 1, 2 etc) as the bam file. Could you 1) check the chr encoding in your bam and fasta files and 2) try to identify the last step that seems to have completed successfully? For example, does the `common_variants_covered.vcf` have variants in it and does the `souporcell_minimap_tagged_sorted.bam` look good as well?

**Comment by mschilli87** [comment]
@drneavin: Just to be on the same page: I ran into that error using the small dataset as well.
However, it looks like your hunch regarding the chromosome encoding was correct: The BAM file provided in the test data uses `1`, `2`, ..., `22` as `RNAME`s, while my genome FASTA uses `chr1`, `chr2`, ..., `chr22` for the same contigs instead. I had chosen that file based on [your response](https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/issues/22#issuecomment-1790403745) to https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/issues/22, where you mentioned `hg38`, which - AFAIK - implies the `chr` prefix. Your addition to [the Data Prep page](https://demultiplexing-doublet-detecting-docs.readthedocs.io/en/latest/DataPrep.html#test-dataset) in response to https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/issues/22 however correctly (AFAIC) refer to `GRCh38` which - AFAIK - is the same assembly as `hg38` but without the `chr` prefix (more precisely: Ensembl rather than UCSC style, but with only autosome involved, the `chr` should be the only difference relevant here).
I'll make sure this fixes my problem and will report back.
If so, do you prefer me to close this as a non-issue or would you be up to convert this into a feature request for adding a quick check for reference assembly compatibility between the BAM and the FASTA files to the documentation? Alternatively, this could be moved upstream as the error message was not particularly helpful but I guess there I'd have to blame Souporcell/troublet rather than Demuxafy. 😉 
Anyways thank you for your fast and helpful response.

**Comment by drneavin** [comment]
Thanks @mschilli87! Yes, I think I probably have a bad habit of using the nomenclature interchangeably and this is a good reminder to tighten it up. I'll accept the pull request shortly. I like the idea of checking for compatibility between the bam and vcf files for softwares that require them since this is a common issue that I think many people run into and can be time consuming and frustrating to track down. I can think of two options to resolve this:

1. I could write a quick script to test for compatibility that anyone could run on their file - this wouldn't be called by each of the methods and would instead rely on the user to call the compatibility script themselves.
2. The other option is that I provide a wrapper for souporcell that would first test the file compatibility and then run souporcell. I would build this into wrappers for the other tools as well. In this case, I could have it as a standalone option and built into how each tool gets run in Demuxafy. 

Do you have thoughts on which would be best from a user standpoint? I think 2 is the better option since it doesn't add another step that could be missed since it would be built-din but will take a little bit more effort in updating the scripts, image and docs to match.

**Comment by mschilli87** [comment]
I like the idea of incorporating it into the wrappers as well.

**Comment by drneavin** [comment]
Ok, great. I can definitely make that happen but it unfortunately won't be a priority for at least a few weeks while I have other projects ongoing. But I'll put it on my to-do list and incorporate it into the next release. I'll leave this open as a reminder. Thanks so much for the suggestions!

**Comment by mschilli87** [comment]
@drneavin: Can this be closed now with https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/commit/de71c31cfcf7dd927d5b2c258a8dd163f3de6001?

**Comment by drneavin** [comment]
I'm keeping the issues open until I can finalise the singularity image and issue a new release (it's always a challenge to rebuild the singularity image since there are often competing dependencies so takes a bit of time to sort out). Once I finalize it and am able to release it I'll close this and the other issues that this new update has addressed

**Comment by drneavin** [comment]
Now released [v3.0.0](https://github.com/drneavin/Demultiplexing_Doublet_Detecting_Docs/releases/tag/v3.0.0)!
